### PR TITLE
Flavor Text returns

### PR DESCRIPTION
**Pull request by Cenrus**
## About The Pull Request

Reimplements flavor text. OOC notes not included

## Why It's Good For The Game

It's hard to express your character with the ss13 sprite limitations so some short sentences can do wonders.

## Changelog

:cl:
add: Flavor text reintroduced
/:cl:

### CODE REVIEW

File: code/modules/mob/living/carbon/human/examine.dm [hunk 1]
```diff
@@ -343,6 +343,19 @@
 	else if(isobserver(user))
 		. += span_info("<b>Traits:</b> [get_quirk_string(FALSE, CAT_QUIRK_ALL)]")
 
+	var/flavor_text_link
+	/// The first 1-FLAVOR_PREVIEW_LIMIT characters in the mob's "flavor_text" DNA feature. FLAVOR_PREVIEW_LIMIT is defined in flavor_defines.dm.
+	var/preview_text = copytext_char((dna.features["flavor_text"]), 1, FLAVOR_PREVIEW_LIMIT)
+	// What examine_tgui.dm uses to determine if flavor text appears as "Obscured".
+	var/face_obscured = (wear_mask && (wear_mask.flags_inv & HIDEFACE)) || (head && (head.flags_inv & HIDEFACE))
```
Comment (Kapu1178) [inline code/modules/mob/living/carbon/human/examine.dm:350]: This is already a var assigned earlier in the proc, `skip_face` (or was it `skipface`?)

File: code/modules/mob/living/carbon/human/examine.dm [hunk 2]
```diff
@@ -343,6 +343,19 @@
 	else if(isobserver(user))
 		. += span_info("<b>Traits:</b> [get_quirk_string(FALSE, CAT_QUIRK_ALL)]")
 
+	var/flavor_text_link
+	/// The first 1-FLAVOR_PREVIEW_LIMIT characters in the mob's "flavor_text" DNA feature. FLAVOR_PREVIEW_LIMIT is defined in flavor_defines.dm.
+	var/preview_text = copytext_char((dna.features["flavor_text"]), 1, FLAVOR_PREVIEW_LIMIT)
```
Comment (Kapu1178) [inline code/modules/mob/living/carbon/human/examine.dm:348]: Really not a fan of this being stored in DNA.features.

File: code/modules/mob/living/carbon/human/examine.dm [hunk 3]
```diff
@@ -343,6 +343,19 @@
...
+	// What examine_tgui.dm uses to determine if flavor text appears as "Obscured".
+	var/face_obscured = (wear_mask && (wear_mask.flags_inv & HIDEFACE)) || (head && (head.flags_inv & HIDEFACE))
+	if(preview_text)
+		if (!(face_obscured))
+			flavor_text_link = span_notice("[preview_text]... <a href='?src=[REF(src)];open_examine_panel=1'>Look closer?</a>")
```
Comment (Kapu1178) [inline code/modules/mob/living/carbon/human/examine.dm:353]: button_element() is a helper for creating hyperlinks

File: code/modules/mob/living/carbon/human/human.dm [hunk 4]
```diff
@@ -102,6 +102,10 @@
 		if(check_obscured_slots(TRUE) & slot)
 			to_chat(usr, span_warning("You can't reach that! Something is covering it."))
 			return
+	if(href_list["open_examine_panel"])
+		var/content = dna.features["flavor_text"]
+		usr << browse("<HTML><HEAD><meta http-equiv='Content-Type' content='text/html; charset=UTF-8'><TITLE>[name]</TITLE></HEAD><BODY><TT>[replacetext(content, "\n", "<BR>")]</TT></BODY></HTML>", "window=[name];size=500x200")
+		onclose(usr, "[name]")
```
Comment (Kapu1178) [inline code/modules/mob/living/carbon/human/human.dm:108]: Use a browser datum for this. Examples can be found all over, but short:
```
var/datum/browser/popup = new(usr, "examine-[REF(src)]", name, 500, 200)
popup.set_content(content)
popup.open(usr)
```

File: code/modules/client/preferences/flavortext.dm [hunk 5]
```diff
@@ -2,5 +2,22 @@
...
+
+/datum/preference/text/flavor_text/user_edit(mob/user, datum/preferences/prefs)
+	var/input = tgui_input_text(user, "Describe your character in greater detail.",, serialize(prefs.read_preference(type)))
+	if(!input)
+		return
+	. = prefs.update_preference(src, input)
+
+	if(istype(user, /mob/dead/new_player))
+		var/mob/dead/new_player/player = user
+		player.new_player_panel()
```
Comment (Kapu1178) [inline code/modules/client/preferences/flavortext.dm:18]: This is rather odd, what's it for?

File: code/modules/client/preferences/flavortext.dm [hunk 6]
```diff
@@ -2,5 +2,22 @@
 	savefile_identifier = PREFERENCE_CHARACTER
 	savefile_key = "flavor_text"
 
+	explanation = "Flavor Text"
+
+/datum/preference/text/flavor_text/get_button(datum/preferences/prefs)
+	return button_element(prefs, "Set Examine Text", "pref_act=[type]")
+
+/datum/preference/text/flavor_text/user_edit(mob/user, datum/preferences/prefs)
+	var/input = tgui_input_text(user, "Describe your character in greater detail.",, serialize(prefs.read_preference(type)))
```
Comment (Kapu1178) [inline code/modules/client/preferences/flavortext.dm:11]: Needs a sensible length limit.

**Review by Kapu1178 — CHANGES_REQUESTED** [review]

File: code/modules/mob/living/carbon/human/examine.dm (hunk 2 above)
Comment (gonenoculer5) [reply code/modules/mob/living/carbon/human/examine.dm:348]: As long as storing it elsewhere does not replicate the problem I have seen in other servers where when the client disconnects, their flavor text doesnt go with the client which causes weird issues. @Kapu1178

**Comment by gonenoculer5** [comment]
An additional safety measure, should a flavortext contain a link, the link (and associated user and character) needs to be logged in ADMINLOG, so that the content of the link can be verified. Might need to work with Kapu on that. I am requesting this for the sake of administrations sanity. I will be writing policy regarding Flavortext and passing it for administration and the leads approval before this is merged.

**Comment by gonenoculer5** [comment]
> An additional safety measure, should a flavortext contain a link, the link (and associated user and character) needs to be logged in ADMINLOG, so that the content of the link can be verified. Might need to work with Kapu on that. I am requesting this for the sake of administrations sanity. I will be writing policy regarding Flavortext and passing it for administration and the leads approval before this is merged.

@Kapu1178 @francinum @fighterslam

**Comment by Kapu1178** [comment]
> An additional safety measure, should a flavortext contain a link, the link (and associated user and character) needs to be logged in ADMINLOG, so that the content of the link can be verified. Might need to work with Kapu on that. I am requesting this for the sake of administrations sanity. I will be writing policy regarding Flavortext and passing it for administration and the leads approval before this is merged.

They won't embed. Embedding player-written HTML is never intentional. It is no different than a player putting a link in papercode, or in chat.

**Comment by gonenoculer5** [comment]
> 

Thats fine. I dont want it to embed, just have the link logged so we can check it.

**Comment by Kapu1178** [comment]
> > 
> 
> Thats fine. I dont want it to embed, just have the link logged so we can check it.

I don't think thats necessary at all. it is the same as papercode/chat.

**Comment by gonenoculer5** [comment]
After discussing with the leads, we'll just make links against the rules for flavortext. 4-6k characters is fine for the text limit.

**Comment by Kapu1178** [comment]
> After discussing with the leads, we'll just make links against the rules for flavortext. 4-6k characters is fine for the text limit.

fuck no 4-6k is INSANE, people do not need to be writing autobiographies

**Comment by Cenrus** [comment]
`MAX_FLAVOR_LEN` is at 4096 at the moment, I haven't changed the value myself

File: code/modules/mob/living/carbon/human/examine.dm (hunk 2 above)
Comment (RimiNosha) [reply code/modules/mob/living/carbon/human/examine.dm:348]: May be best to store it in mind instead.

**Comment by Kapu1178** [comment]
> `MAX_FLAVOR_LEN` is at 4096 at the moment, I haven't changed the value myself

Ill deal with it later, dw

**Comment by Kapu1178** [comment]
@Cenrus Please address remaining reviews

**Pull request merged**